### PR TITLE
Bytewise crc32

### PR DESCRIPTION
**Pull request by k-danil**
Good day!

This pull request introduce bytewise crc32 calculation instead of current bitwise. Code based on VLC implementation, but instead per-object table generation, it use static crc32 table (+1kb additional lifetime memory use but no recalculations and reallocations) .
Current:
```
Benchmark_updateCRC32/Calc_PAT_crc32
Benchmark_updateCRC32/Calc_PAT_crc32-8         	 9470034	       122.9 ns/op	       0 B/op	       0 allocs/op
Benchmark_updateCRC32/Calc_PMT_crc32
Benchmark_updateCRC32/Calc_PMT_crc32-8         	 4010506	       283.4 ns/op	       0 B/op	       0 allocs/op
```
Introduced in this patch:
```
Benchmark_updateCRC32/Calc_PAT_crc32_precomputed
Benchmark_updateCRC32/Calc_PAT_crc32_precomputed-8         	71367823	        16.79 ns/op	       0 B/op	       0 allocs/op
Benchmark_updateCRC32/Calc_PMT_crc32_precomputed
Benchmark_updateCRC32/Calc_PMT_crc32_precomputed-8         	18918052	        54.75 ns/op	       0 B/op	       0 allocs/op
```

Test's and benchmarks based on real TS headers. Old implementation renamed and kept in file for comparison sake, you can remove it safely.

### CODE REVIEW

File: crc32.go [hunk 1]
```diff
@@ -4,13 +4,88 @@ const (
...
+		}
+		fmt.Printf("%#X", k)
+	}
+	fmt.Println(`}`)
+}
+
+*/
+// https://github.com/videolan/vlc/blob/master/modules/mux/mpeg/ps.c
+
+var tableCRC32 = [256]uint32{
```
Comment (asticode) [inline crc32.go:36]: Could you create a command to generate this table in a separate go file?

1) Add an `internal/cmd/crc32_table` folder
2) Add a `main.go` in it that would create a `crc32_table.go` with only the `tableCRC32` variable in it.
3) Also make sure to add `// Code generated by astits. DO NOT EDIT.` a the top of this file before the package name

That way we have a command to generate the file

File: crc32.go [hunk 2]
```diff
@@ -4,13 +4,88 @@ const (
...
+func updateCRC32(iCrc uint32, bs []byte) uint32 {
+	for _, b := range bs {
+		iCrc = (iCrc << 8) ^ tableCRC32[((iCrc>>24)^uint32(b))&0xff]
+	}
+	return iCrc
+}
+
+// computeCRC32 computes a CRC32
+// https://stackoverflow.com/questions/35034042/how-to-calculate-crc32-in-psi-si-packet
+func computeCRC32old(bs []byte) uint32 {
```
Comment (asticode) [inline crc32.go:84]: You can remove the old implementation

File: crc32_test.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+		crc:  binary.BigEndian.Uint32(testDataPat[len(testDataPat)-4:]),
+		data: testDataPat[:len(testDataPat)-4],
+	}, {
+		name: "Calc PMT crc32",
+		crc:  binary.BigEndian.Uint32(testDataPmt[len(testDataPmt)-4:]),
+		data: testDataPmt[:len(testDataPmt)-4],
+	},
+}
+
+func Benchmark_updateCRC32(b *testing.B) {
```
Comment (asticode) [inline crc32_test.go:31]: No need to keep the benchmark here

File: crc32_test.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+					panic("incorrect sum")
+				}
+			}
+		})
+	}
+}
+
+func Test_updateCRC32(t *testing.T) {
+
+	for _, test := range tests {
```
Comment (asticode) [inline crc32_test.go:61]: You can remove this

File: crc32_test.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+		t.Run(test.name, func(t *testing.T) {
+			assert.Equal(t, test.crc, computeCRC32old(test.data))
+		})
+	}
+
+	for _, test := range tests {
+		t.Run(test.name+" precomputed", func(t *testing.T) {
```
Comment (asticode) [inline crc32_test.go:68]: You can remove the " precomputed" here

**Review by asticode — CHANGES_REQUESTED** [review]
Nice PR!

I'd rather see it organized differently, I've made comments.

File: crc32.go (hunk 1 above)
Comment (k-danil) [reply crc32.go]: All done! This is my first attempt with code generation, so it may be kind of ugly

File: crc32.go (hunk 2 above)
Comment (k-danil) [reply crc32.go]: Done

File: crc32_test.go (hunk 3 above)
Comment (k-danil) [reply crc32_test.go]: Done

File: crc32_test.go (hunk 4 above)
Comment (k-danil) [reply crc32_test.go]: Done

File: crc32_test.go (hunk 5 above)
Comment (k-danil) [reply crc32_test.go]: Done

File: crc32.go [hunk 6]
```diff
@@ -4,22 +4,17 @@ const (
 	crc32Polynomial = uint32(0xffffffff)
 )
 
-// computeCRC32 computes a CRC32
-// https://stackoverflow.com/questions/35034042/how-to-calculate-crc32-in-psi-si-packet
+// this solution based on vlc implementation + static crc table (1kb additional memory on start, without reallocations)
```
Comment (asticode) [inline crc32.go:7]: Could you put this comment above `updateCRC32` instead ?

And could you replace it with : 

```
// Based on VLC implementation using a static CRC table (1kb additional memory on start, without 
// reallocations): https://github.com/videolan/vlc/blob/master/modules/mux/mpeg/ps.c
```

File: internal/cmd/crc32_table/main.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+const (
+	disclaimer = `// Code generated by astits. DO NOT EDIT`
```
Comment (asticode) [inline internal/cmd/crc32_table/main.go:10]: Could you update the disclaimer to `// Code generated by astits using internal/cmd/crc32_table. DO NOT EDIT`

File: internal/cmd/crc32_table/main.go [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+	packet     = `package astits`
+	filename   = `crc32_table.go`
+)
+
+func main() {
+	file, err := os.OpenFile(filename, os.O_WRONLY|os.O_CREATE|os.O_TRUNC|os.O_SYNC, 0644)
```
Comment (asticode) [inline internal/cmd/crc32_table/main.go:16]: Could you use `os.Create()` instead ?

File: internal/cmd/crc32_table/main.go [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+	if err != nil {
+		panic(err)
```
Comment (asticode) [inline internal/cmd/crc32_table/main.go:18]: Could you use `log.Fatal(err)` instead ?

File: internal/cmd/crc32_table/main.go [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+		if i%8 == 0 {
+			doOrPanic(fmt.Fprintf(w, "\n\t"))
+		}
+		doOrPanic(fmt.Fprintf(w, "%#08X, ", k))
+	}
+	doOrPanic(fmt.Fprintf(w, "\n%s", `}`))
+}
+
+func doOrPanic(_ int, err error) {
```
Comment (asticode) [inline internal/cmd/crc32_table/main.go:47]: Could you rename this to `write()` and instead of `panic(err)` do `log.Fatal(err)` ?

**Review by asticode — CHANGES_REQUESTED** [review]

File: crc32.go (hunk 6 above)
Comment (k-danil) [reply crc32.go]: Done

File: internal/cmd/crc32_table/main.go (hunk 8 above)
Comment (k-danil) [reply internal/cmd/crc32_table/main.go]: Yeah, this approach is much neater, thanks!

File: internal/cmd/crc32_table/main.go (hunk 9 above)
Comment (k-danil) [reply internal/cmd/crc32_table/main.go]: Done

File: internal/cmd/crc32_table/main.go (hunk 10 above)
Comment (k-danil) [reply internal/cmd/crc32_table/main.go]: On my opinion it is more general purpose function then just writer (and it's not even write, just crash if something go wrong). Maybe 'try()' will be more appropriate name? Anyway, i done as you ask

**Comment by asticode** [comment]
fyi I've created a `v1.11.0` tag

thanks for the PR! ❤️

**Pull request merged**